folsom-opentelemetry-metrics/src/test/java/com/spotify/folsom/client/OpenTelemetryMetricsTest.java

```diff
@@ -0,0 +1,543 @@
+        metrics.registerOutstandingRequestsGauge(gauge2);
+        metrics.registerOutstandingRequestsGauge(gauge3);
+
+        metrics.unregisterOutstandingRequestsGauge(gauge1);
+        metrics.unregisterOutstandingRequestsGauge(gauge3);
+
+        // then
+        GaugeData<DoublePointData> gaugeData = getGaugeByName("memcache.outstanding.requests");
+        DoublePointData doublePointData = new ArrayList<>(gaugeData.getPoints()).get(0);
+
+        assertEquals(321.0, doublePointData.getValue(), EPS);
+    }
+
+    @Test
+    public void shouldRegisterGaugesWithoutRaceCondition() throws InterruptedException {
+        // given
+        int numThreads = 10;
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        CountDownLatch latch = new CountDownLatch(numThreads);
+
+        // when
+        for (int i = 0; i < numThreads; i++) {
+            executorService.submit(() -> {
+                for (int j = 1; j <= 999; j++) {
+                    int outstandingRequests = j;
+                    metrics.registerOutstandingRequestsGauge(() -> outstandingRequests);
+                }
+                latch.countDown();
+            });
+        }
+
+        latch.await();
+        executorService.shutdown();
+
+        // then
+        GaugeData<DoublePointData> gaugeData = getGaugeByName("memcache.outstanding.requests");
+        DoublePointData doublePointData = new ArrayList<>(gaugeData.getPoints()).get(0);
+        double nth_partial_sum = (double) (999 * 1000) / 2; // https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_%E2%8B%AF
+        double sum = nth_partial_sum * numThreads;
+
+        assertEquals(sum, doublePointData.getValue(), EPS);
+    }
+
+    private GaugeData<DoublePointData> getGaugeByName(String gaugeName) {
+        return otelTesting.getMetrics()
+                .stream()
+                .filter(metricData -> Objects.equals(gaugeName, metricData.getName()))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("gauge not found"))
+                .getDoubleGaugeData();
+    }
+
+    private <T extends PointData> T findMetric(String metricName,
+                                               String key,
+                                               String value,
+                                               Class<T> targetClass) {
+        PointData pointData = otelTesting.getMetrics()
+                .stream()
+                .filter(metricData -> Objects.equals(metricName, metricData.getName()))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("metric not found"))
+                .getData()
+                .getPoints()
+                .stream()
+                .filter(pd -> Objects.equals(value, pd.getAttributes().get(AttributeKey.stringKey(key))))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("attr not found"));
```

Comment:
If you're fine with using AssertJ, the `sdk-testing` module provides some utilities & assertion classes, see https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/testing/src/main/java/io/opentelemetry/sdk/testing/assertj/OpenTelemetryAssertions.java